x-pack/plugins/fields_metadata/common/hashed_cache.ts

```diff
@@ -0,0 +1,46 @@
+/*
```

Comment:
I suspect this service would be okay without a cache UNLESS someone starts loading the whole field list in one go.

Relying on the browser http cache would be a bit more effective than an in memory cache. I think it would be a little bit more work, but hopefully not a ton. 

Hm, now I'm realizing that I'm not certain if this is caching requests or individual fields data - an important detail.

Reply:
> Hm, now I'm realizing that I'm not certain if this is caching requests or individual fields data - an important detail.

To clarify, the cache is for the response data based on the search params, not on an individual field level, but it's something we can think of in future, although it might overcomplicate the implementation now without a real need.

> I suspect this service would be okay without a cache UNLESS someone starts loading the whole field list in one go.

That might be the case as @kertal already asked about in [this comment](https://github.com/elastic/kibana/pull/183806#pullrequestreview-2077559042).

I feel the cache at this client level might be good for various reasons. 

Say we want to show individual field descriptions in the sidebar on Discover and we locate a custom hook call for the field metadata at the field level, so that the request is triggered only when a field is hovered (just thinking of a use case, doesn't need to be like that).

In case the data source is changed, it would trigger a full rerender of the sidebar, while many fields would remain the same. In this case the client cache is great to keep those metadata available for the same field and avoid a round-trip to the server.

> Relying on the browser http cache would be a bit more effective than an in memory cache.

I agree, the native browser cache is surely effective when it comes to manage heavy-load requests and assets loading.
I opted for a faster solution using the in-memory cache for a couple of reasons:
- As you mentioned, that would require a bit more work although not a huge effort, I preferred to keep the scope of this work small since it already is a big PR.
- HTTP browser caching would preclude what mentioned on another comment about having a more granular cache on individual fields, so it might be worth to also keep the door open to this possibility.

This is anyway worth revisiting in future, my recommendation is to keep it simple for now and iterate where we need enhancements for upcoming use cases.

Let me know your thoughts and thanks for raising this point!

Reply:
> > Hm, now I'm realizing that I'm not certain if this is caching requests or individual fields data - an important detail.
> 
> To clarify, the cache is for the response data based on the search params, not on an individual field level, but it's something we can think of in future, although it might overcomplicate the implementation now without a real need.

just wanted to ask if you clean up the cache, seeing you're using LRU cache, nice 👍 

> 
> > I suspect this service would be okay without a cache UNLESS someone starts loading the whole field list in one go.
> 
> That might be the case as @kertal already asked about in [this comment](https://github.com/elastic/kibana/pull/183806#pullrequestreview-2077559042).
> 
> I feel the cache at this client level might be good for various reasons.
> 
> Say we want to show individual field descriptions in the sidebar on Discover and we locate a custom hook call for the field metadata at the field level, so that the request is triggered only when a field is hovered (just thinking of a use case, doesn't need to be like that).

yes, exactly, question is, if we need to cache for this use case, when just a single field is requested because of the description. this might not be worth the memory we're using for the cache in this case. The more fields are requested, the more important caching is. So e.g. for the ES|QL and related use case, where we need to determine which field is ECS and which not, and where I think we might load the whole field list upfront, so it's always snappy? (Question to @elastic/kibana-esql)

I think @mattkime's suggestion to cache on the http layer is definitely worth pursuing, we had great experience with the recent improvements  when starting to use the  stale-while-revalidate strategy for data view field caps, but yes, this could be done as a future improvement.

Reply:
Thanks for sharing, I remember this piece of work and we might not need to cover all the same behaviours (maybe `stale-while-revalidate` is not essential for this usage) but I'll take a look!

Do you consider this part a blocker for the PR to be merged? Under a features perspective there is not additional benefit apart the technical implementation, and we have an upcoming OnWeek where I could take some time to look at this as follow-up work.

Reply:
>Do you consider this part a blocker for the PR to be merged?

Not at all.

I think the primary value is getting more devs thinking about the browser cache.